From 7f97c7537cb4cfc53b6ae00a3d4292586f42f213 Mon Sep 17 00:00:00 2001
From: OpenEmbedded <oe.patch@oe>
Date: Fri, 29 May 2020 15:04:08 +0200
Subject: [PATCH] patched usbduxfast module

---
 drivers/staging/comedi/drivers/usbduxfast.c | 324 +++++++++++++++-----
 1 file changed, 254 insertions(+), 70 deletions(-)

diff --git a/drivers/staging/comedi/drivers/usbduxfast.c b/drivers/staging/comedi/drivers/usbduxfast.c
index e18c0723b760..cfd7779f5a5a 100644
--- a/drivers/staging/comedi/drivers/usbduxfast.c
+++ b/drivers/staging/comedi/drivers/usbduxfast.c
@@ -1,6 +1,15 @@
-// SPDX-License-Identifier: GPL-2.0+
 /*
  *  Copyright (C) 2004-2014 Bernd Porr, mail@berndporr.me.uk
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
  */
 
 /*
@@ -20,6 +29,8 @@
  *
  * Bernd Porr
  *
+ * Various revisions made to allow full-rate sampling of 16 channels,
+ * copyright (C) 2014 John Hallam (JCTH)
  *
  * Revision history:
  * 0.9: Dropping the first data packet which seems to be from the last transfer.
@@ -91,9 +102,11 @@
 #define SIZEADIN	(sizeof(s16))
 
 /*
- * size of the input-buffer IN BYTES
+ * size of the input-buffer IN USB Bulk Packet Buffers (512)
+ * -- the default for the urbsz module parameter.
  */
-#define SIZEINBUF	512
+#define URBSZINBUF	8
+#define MAX_URBSZINBUF	16	/* Limit max buffer size for kernel's sake! */
 
 /*
  * 16 bytes
@@ -106,9 +119,10 @@
 #define SIZEOFDUXBUF	256
 
 /*
- * number of in-URBs which receive the data: min=5
+ * number of in-URBs which receive the data: MAX and DEFAULT
  */
-#define NUMOFINBUFFERSHIGH	10
+#define MAX_NUM_URBS	 8
+#define DEFAULT_NUM_URBS 4
 
 /*
  * min delay steps for more than one channel
@@ -123,6 +137,18 @@
  */
 #define MAX_SAMPLING_PERIOD	500
 
+/*
+ * Convert FX2 GPIF ticks to nanoseconds and vice versa by integer arithmetic.
+ *
+ * NS_TO_TICKS quantizes the requested delay into 30 MHz clock ticks.
+ * TICKS_TO_NS computes the inter-channel delay given the sampling loop length in ticks.
+ *
+ * It is important that NS_TO_TICKS(TICKS_TO_NS(.)) is the identity function.
+ * N.B. Avoid side-effects in the arguments to these macros.
+ */
+#define NS_TO_TICKS(ns) ((30*(ns)+500)/1000)
+#define TICKS_TO_NS(nt)	(100*((nt)/3) + 33*((nt)%3) + (((nt)%3) == 2))
+
 /*
  * number of received packets to ignore before we start handing data
  * over to comedi, it's quad buffering and we have to ignore 4 packets
@@ -146,14 +172,44 @@ static const struct comedi_lrange range_usbduxfast_ai_range = {
  * one sub device just now: A/D
  */
 struct usbduxfast_private {
-	struct urb *urb;	/* BULK-transfer handling: urb */
-	u8 *duxbuf;
-	s8 *inbuf;
+	struct urb *urb[MAX_NUM_URBS];	/* BULK-transfer handling: urb */
+	s8  *inbuf[MAX_NUM_URBS];
+	int  urb_active[MAX_NUM_URBS];	/* Needed if an error to call kill on URB not submitted */
+	s8  *insnbuf;
+	u8  *duxbuf;
 	short int ai_cmd_running;	/* asynchronous command is running */
 	int ignore;		/* counter which ignores the first buffers */
 	struct mutex mut;
 };
 
+/*
+ * Maximum usable number of URBs:  is MAX_NUM_URBS rounded to even.
+ */
+static const int max_num_urbs = (MAX_NUM_URBS&~1);
+
+/*
+ * Number of URBs allocated per (sub)device.  Default is DEFAULT_NUM_URBS.
+ * Instantiated by the probe function.
+ */
+static int nurbs = 0;
+module_param(nurbs, int, S_IRUGO);
+MODULE_PARM_DESC(nurbs, "Number of InURBs to allocate (default 4)");
+
+/*
+ * Are the URBs paired or not?
+ */
+static int paired = 1;
+module_param(paired, int, S_IRUGO);
+MODULE_PARM_DESC(paired, "Use InURBs in pairs if non-zero (default 1);  implies nurbs is even");
+
+/*
+ * URB request buffer size, as a multiple of 512 bytes
+ * The default is set in the probe function from URBSZINBUFUSB
+ */
+static int urbsz = 0;
+module_param(urbsz, int, S_IRUGO);
+MODULE_PARM_DESC(urbsz, "Size of transfer requested by an InURBs in USB bulk packets (default 8)");
+
 /*
  * bulk transfers to usbduxfast
  */
@@ -198,9 +254,14 @@ static int usbduxfast_ai_stop(struct comedi_device *dev, int do_unlink)
 	/* stop aquistion */
 	devpriv->ai_cmd_running = 0;
 
-	if (do_unlink && devpriv->urb) {
-		/* kill the running transfer */
-		usb_kill_urb(devpriv->urb);
+	if (do_unlink) {
+	  int j;
+	  /* kill the running transfer(s) */
+	  for(j=0; j<nurbs; j++) {
+	    if( devpriv->urb_active[j] )
+	      usb_kill_urb(devpriv->urb[j]);
+	    devpriv->urb_active[j] = 0;
+	  }
 	}
 
 	return 0;
@@ -219,6 +280,21 @@ static int usbduxfast_ai_cancel(struct comedi_device *dev,
 	return ret;
 }
 
+/*
+ * Paired URB helper function -- assumes not too many URBs.
+ */
+
+static int urb_index(struct usbduxfast_private *udfp, struct urb *urb)
+{
+  int i;
+
+  for(i=0; i<max_num_urbs; i++) {
+    if( udfp->urb[i] == urb )
+      return i;
+  }
+  return -1;
+}
+
 static void usbduxfast_ai_handle_urb(struct comedi_device *dev,
 				     struct comedi_subdevice *s,
 				     struct urb *urb)
@@ -226,8 +302,48 @@ static void usbduxfast_ai_handle_urb(struct comedi_device *dev,
 	struct usbduxfast_private *devpriv = dev->private;
 	struct comedi_async *async = s->async;
 	struct comedi_cmd *cmd = &async->cmd;
+	int idx = urb_index(devpriv, urb);
 	int ret;
 
+	if (idx < 0) {
+	    /*
+	     * This should not happen: a URB called back that is not in the list!
+	     */
+	  dev_err(dev->class_dev, "index locate for urb %p failed", urb);
+	  async->events |= COMEDI_CB_EOA;
+	  async->events |= COMEDI_CB_ERROR;
+	  comedi_event(dev, s);
+	  usbduxfast_ai_stop(dev, 0);
+	  return;
+	}
+
+	devpriv->urb_active[idx] = 0;
+
+	/*
+	 * If pairing and command still running, submit the paired urb now, if not resubmit this one at end
+	 */
+	if (paired && !(async->events & COMEDI_CB_CANCEL_MASK)) {
+	  int pair = idx ^ 1;
+	  struct urb *next_urb = devpriv->urb[pair]; /* Locate the pair */
+	  int err = 0;
+
+	  next_urb->dev = comedi_to_usb_dev(dev);
+	  next_urb->status = 0;
+	  err = usb_submit_urb(next_urb, GFP_ATOMIC);
+	  if (err < 0) {
+	    /*
+	     * Fixme: This throws away a valid URB (the current one) if pair submission fails
+	     */
+	    dev_err(dev->class_dev, "paired urb submit failed: %d", err);
+	    async->events |= COMEDI_CB_EOA;
+	    async->events |= COMEDI_CB_ERROR;
+	    comedi_event(dev, s);
+	    usbduxfast_ai_stop(dev, 0);
+	    return;
+	  }
+	  devpriv->urb_active[pair]++;
+	}
+
 	if (devpriv->ignore) {
 		devpriv->ignore--;
 	} else {
@@ -243,14 +359,15 @@ static void usbduxfast_ai_handle_urb(struct comedi_device *dev,
 	}
 
 	/* if command is still running, resubmit urb for BULK transfer */
-	if (!(async->events & COMEDI_CB_CANCEL_MASK)) {
+	if (!(async->events & COMEDI_CB_CANCEL_MASK) && !paired) {
 		urb->dev = comedi_to_usb_dev(dev);
 		urb->status = 0;
 		ret = usb_submit_urb(urb, GFP_ATOMIC);
 		if (ret < 0) {
-			dev_err(dev->class_dev, "urb resubm failed: %d", ret);
+			dev_err(dev->class_dev, "urb resubmit failed: %d", ret);
 			async->events |= COMEDI_CB_ERROR;
 		}
+		devpriv->urb_active[idx]++;
 	}
 }
 
@@ -301,17 +418,25 @@ static int usbduxfast_submit_urb(struct comedi_device *dev)
 {
 	struct usb_device *usb = comedi_to_usb_dev(dev);
 	struct usbduxfast_private *devpriv = dev->private;
-	int ret;
+	const int InURB_buf_size = 512 * urbsz;
+	int j, ret;
 
-	usb_fill_bulk_urb(devpriv->urb, usb, usb_rcvbulkpipe(usb, BULKINEP),
-			  devpriv->inbuf, SIZEINBUF,
-			  usbduxfast_ai_interrupt, dev);
+	for(j=0; j<nurbs && devpriv->urb[j]; j++) {
+	  usb_fill_bulk_urb(devpriv->urb[j], usb, usb_rcvbulkpipe(usb, BULKINEP),
+			    devpriv->inbuf[j], InURB_buf_size,
+			    usbduxfast_ai_interrupt, dev);
+	}
 
-	ret = usb_submit_urb(devpriv->urb, GFP_ATOMIC);
-	if (ret) {
-		dev_err(dev->class_dev, "usb_submit_urb error %d\n", ret);
-		return ret;
+	for(j=0; j<nurbs && devpriv->urb[j]; j++) {
+	  ret = usb_submit_urb(devpriv->urb[j], GFP_ATOMIC);
+	  if (ret) {
+	    dev_err(dev->class_dev, "usb_submit_urb[%d] error %d\n", j, ret);
+	    return ret;
+	  }
+	  devpriv->urb_active[j]++;
+	  if (paired) j++;	/* Only submit the first of each pair now */
 	}
+
 	return 0;
 }
 
@@ -397,12 +522,13 @@ static int usbduxfast_ai_cmdtest(struct comedi_device *dev,
 	 * for all other chanlist_len it is:
 	 *	MIN_SAMPLING_PERIOD <= steps <= MAX_SAMPLING_PERIOD
 	 */
-	steps = (cmd->convert_arg * 30) / 1000;
+	steps = NS_TO_TICKS(cmd->convert_arg);
 	if (cmd->chanlist_len !=  1)
 		err |= comedi_check_trigger_arg_min(&steps,
 						    MIN_SAMPLING_PERIOD);
 	err |= comedi_check_trigger_arg_max(&steps, MAX_SAMPLING_PERIOD);
-	arg = (steps * 1000) / 30;
+	/* calc arg again -- correct for tick quantisation */
+	arg = TICKS_TO_NS(steps);
 	err |= comedi_check_trigger_arg_is(&cmd->convert_arg, arg);
 
 	if (cmd->stop_src == TRIG_COUNT)
@@ -475,7 +601,7 @@ static int usbduxfast_ai_cmd(struct comedi_device *dev,
 	 */
 	devpriv->ignore = PACKETS_TO_IGNORE;
 
-	steps = (cmd->convert_arg * 30) / 1000;
+	steps = NS_TO_TICKS(cmd->convert_arg);
 
 	switch (cmd->chanlist_len) {
 	case 1:
@@ -672,52 +798,60 @@ static int usbduxfast_ai_cmd(struct comedi_device *dev,
 		break;
 
 	case 16:
-		if (CR_RANGE(cmd->chanlist[0]) > 0)
-			rngmask = 0xff - 0x04;
-		else
-			rngmask = 0xff;
+		if (cmd->start_src == TRIG_EXT) {	     /* Trigger on RDY0 state */
 
-		if (cmd->start_src == TRIG_EXT) {
-			/*
-			 * we loop here until ready has been set
-			 */
+		        if( !cmd->start_arg ) {		     /* Trigger is active low */
 
-			/* branch back to state 0 */
-			/* deceision state w/o data */
-			/* reset */
-			/* RDY0 = 0 */
-			usbduxfast_cmd_data(dev, 0, 0x01, 0x01,
-					    (0xff - 0x02) & rngmask, 0x00);
-		} else {
-			/*
-			 * we just proceed to state 1
-			 */
+			  /* State 0 */
+			  /* branch back to state 0 while RDY0 = 1 */
+			  /* decision state w/o data */
+			  /* reset CTL1 is low while waiting, min 33ns */
+			  usbduxfast_cmd_data(dev, 0, 0x10, 0x01,
+					      (0xff - 0x02) & rngmask, 0x00);
+
+			} else {			     /* Trigger is active high */
 
-			/* 30us reset pulse */
-			/* reset */
-			usbduxfast_cmd_data(dev, 0, 0xff, 0x00,
-					    (0xff - 0x02) & rngmask, 0x00);
+			  /* State 0 */
+			  /* branch back to state 0 while RDY0 = 0 */
+			  /* decision state w/o data */
+			  /* reset CTL1 is low while waiting, min 33ns */
+			  usbduxfast_cmd_data(dev, 0, 0x02, 0x01,
+					      (0xff - 0x02) & rngmask, 0x00);
+			}
+		} else {
+			  /* State 0 */
+			  /* branch to state 2 unconditionally */ 
+			  /* 33ns reset pulse */
+			  /* reset CTL1 is low */
+			  usbduxfast_cmd_data(dev, 0, 0x12, 0x01,
+					      (0xff - 0x02) & rngmask, 0x00);
 		}
 
+		/* State 1 */
 		/* commit data to the FIFO */
-		/* data */
-		usbduxfast_cmd_data(dev, 1, 0x01, 0x02, rngmask, 0x00);
+		/* data to FIFO */
+		/* muxclock CTL0 goes high */
+		usbduxfast_cmd_data(dev, 1, 0x01, 0x02, 0xff & rngmask, 0x00);
 
 		/* we have 2 states with duration 1 */
 		steps = steps - 2;
 
+		/* State 2 */
 		/* do the first part of the delay */
 		usbduxfast_cmd_data(dev, 2, steps / 2,
-				    0x00, 0xfe & rngmask, 0x00);
+				    0x00, 0xff & rngmask, 0x00);
 
+		/* State 3 */
 		/* and the second part */
 		usbduxfast_cmd_data(dev, 3, steps - steps / 2,
-				    0x00, rngmask, 0x00);
+				    0x00, 0xff & rngmask, 0x00);
 
+		/* State 4 */
 		/* branch back to state 1 */
-		/* deceision state w/o data */
-		/* doesn't matter */
-		usbduxfast_cmd_data(dev, 4, 0x09, 0x01, rngmask, 0xff);
+		/* decision state w/o data */
+		/* muxclock CTL0 goes low */
+		usbduxfast_cmd_data(dev, 4, 0x09, 0x01,
+				    (0xff - 0x01) & rngmask, 0xff);
 
 		break;
 	}
@@ -796,7 +930,7 @@ static int usbduxfast_ai_insn_read(struct comedi_device *dev,
 
 	for (i = 0; i < PACKETS_TO_IGNORE; i++) {
 		ret = usb_bulk_msg(usb, usb_rcvbulkpipe(usb, BULKINEP),
-				   devpriv->inbuf, SIZEINBUF,
+				   devpriv->insnbuf, SIZEINSNBUF,
 				   &actual_length, 10000);
 		if (ret < 0) {
 			dev_err(dev->class_dev, "insn timeout, no data\n");
@@ -807,7 +941,7 @@ static int usbduxfast_ai_insn_read(struct comedi_device *dev,
 
 	for (i = 0; i < insn->n;) {
 		ret = usb_bulk_msg(usb, usb_rcvbulkpipe(usb, BULKINEP),
-				   devpriv->inbuf, SIZEINBUF,
+				   devpriv->insnbuf, SIZEINSNBUF,
 				   &actual_length, 10000);
 		if (ret < 0) {
 			dev_err(dev->class_dev, "insn data error: %d\n", ret);
@@ -821,7 +955,7 @@ static int usbduxfast_ai_insn_read(struct comedi_device *dev,
 			return -EINVAL;
 		}
 		for (j = chan; (j < n) && (i < insn->n); j = j + 16) {
-			data[i] = ((u16 *)(devpriv->inbuf))[j];
+			data[i] = ((u16 *)(devpriv->insnbuf))[j];
 			i++;
 		}
 	}
@@ -909,7 +1043,8 @@ static int usbduxfast_auto_attach(struct comedi_device *dev,
 	struct usb_device *usb = comedi_to_usb_dev(dev);
 	struct usbduxfast_private *devpriv;
 	struct comedi_subdevice *s;
-	int ret;
+	const  int InURB_buf_size = 512 * urbsz;
+	int j, ret;
 
 	if (usb->speed != USB_SPEED_HIGH) {
 		dev_err(dev->class_dev,
@@ -936,13 +1071,24 @@ static int usbduxfast_auto_attach(struct comedi_device *dev,
 		return -ENODEV;
 	}
 
-	devpriv->urb = usb_alloc_urb(0, GFP_KERNEL);
-	if (!devpriv->urb)
-		return -ENOMEM;
+	for(j=0; j<nurbs; j++) {
+	  devpriv->urb[j] = usb_alloc_urb(0, GFP_KERNEL);
+	  if (!devpriv->urb[j]) {
+	    dev_err(dev->class_dev, "Could not alloc. urb[%d]\n", j);
+	    return -ENOMEM;
+	  }
+	  devpriv->urb_active[j] = 0;
+	}
 
-	devpriv->inbuf = kmalloc(SIZEINBUF, GFP_KERNEL);
-	if (!devpriv->inbuf)
-		return -ENOMEM;
+	for(j=0; j<nurbs; j++) {
+	  devpriv->inbuf[j] = kmalloc(InURB_buf_size, GFP_KERNEL);
+	  if (!devpriv->inbuf[j])
+	    return -ENOMEM;
+	}
+
+	devpriv->insnbuf = kmalloc(SIZEINSNBUF, GFP_KERNEL);
+	if (!devpriv->insnbuf)
+	  return -ENOMEM;
 
 	ret = comedi_load_firmware(dev, &usb->dev, FIRMWARE,
 				   usbduxfast_upload_firmware, 0);
@@ -982,14 +1128,27 @@ static void usbduxfast_detach(struct comedi_device *dev)
 
 	usb_set_intfdata(intf, NULL);
 
-	if (devpriv->urb) {
-		/* waits until a running transfer is over */
-		usb_kill_urb(devpriv->urb);
+	if (devpriv->urb[0]) {
+	  int j;
+
+	  /* waits until a running transfer is over */
+	  for(j=0; j<nurbs && devpriv->urb[j]; j++)
+	    usb_kill_urb(devpriv->urb[j]);
+
+	  for(j=0; j<nurbs && devpriv->urb[j]; j++) {
+	    if( devpriv->inbuf[j] )
+	      kfree(devpriv->inbuf[j]);
+	    devpriv->inbuf[j] = NULL;
 
-		kfree(devpriv->inbuf);
-		usb_free_urb(devpriv->urb);
+	    if( devpriv->urb[j] )
+	      usb_free_urb(devpriv->urb[j]);
+	    devpriv->urb[j] = NULL;
+	    devpriv->urb_active[j] = 0;
+	  }
 	}
 
+	kfree(devpriv->insnbuf);
+	
 	kfree(devpriv->duxbuf);
 
 	mutex_unlock(&devpriv->mut);
@@ -1005,6 +1164,31 @@ static struct comedi_driver usbduxfast_driver = {
 static int usbduxfast_usb_probe(struct usb_interface *intf,
 				const struct usb_device_id *id)
 {
+	if (nurbs <= 0) {
+	  nurbs = DEFAULT_NUM_URBS;
+	}
+
+	if (paired) {
+	  nurbs = (nurbs+1) & ~1; /* Make even, rounding up */
+	}
+
+	if (nurbs > max_num_urbs) {
+	  printk(KERN_WARNING "usbduxfast module: nurbs %d exceeds compiled maximum %d\n", nurbs, max_num_urbs);
+	  nurbs = max_num_urbs;
+	}
+
+	if (urbsz <= 0) {
+	  urbsz = URBSZINBUF;
+	}
+
+	if (urbsz > MAX_URBSZINBUF) {
+	  printk(KERN_WARNING "usbduxfast module: urbsz %d exceeds compiled maximum %d\n", urbsz, MAX_URBSZINBUF);
+	  urbsz = URBSZINBUF;
+	}
+
+	printk(KERN_INFO "usbduxfast module: using %d %s URB(s) of size %d*512 bytes for bulk transfer\n",
+	       nurbs, (paired? "paired" : ""), urbsz); 
+
 	return comedi_usb_auto_config(intf, &usbduxfast_driver, 0);
 }
 
@@ -1024,7 +1208,7 @@ static struct usb_driver usbduxfast_usb_driver = {
 };
 module_comedi_usb_driver(usbduxfast_driver, usbduxfast_usb_driver);
 
-MODULE_AUTHOR("Bernd Porr, BerndPorr@f2s.com");
-MODULE_DESCRIPTION("USB-DUXfast, BerndPorr@f2s.com");
+MODULE_AUTHOR("Bernd Porr, BerndPorr@f2s.com, John Hallam sw@j.hallam.dk");
+MODULE_DESCRIPTION("USB-DUXfast, BerndPorr@f2s.com, John Hallam sw@j.hallam.dk");
 MODULE_LICENSE("GPL");
 MODULE_FIRMWARE(FIRMWARE);
